kube-runtime/src/cache/mod.rs

```diff
@@ -25,6 +29,87 @@ where
     stream.inspect_ok(move |event| store.apply_watcher_event(event))
 }
 
+/// A simple reflector cache around a store and an owned watcher
+pub struct Cache<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash,
+{
+    reader: Store<K>,
+    cache: BoxStream<'static, std::result::Result<watcher::Event<K>, watcher::Error>>,
+}
+
+impl<K> Cache<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash + Clone + Default,
+{
+    /// Create a Cache on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Cache` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    #[must_use]
+    pub fn new(api: Api<K>, lp: ListParams) -> Self {
+        Self::new_with(api, lp, Default::default())
+    }
+}
+
+impl<K> Cache<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash + Clone,
+{
+    /// Create a Cache on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Cache` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    ///
+    /// This variant constructor is for [`dynamic`] types found through discovery. Prefer [`Cache::new`] for static types.
+    #[must_use]
+    pub fn new_with(api: Api<K>, lp: ListParams, dyntype: K::DynamicType) -> Self {
+        let writer = Writer::<K>::new(dyntype);
+        let reader = writer.as_reader();
+        let cache = reflector(writer, watcher(api, lp)).boxed();
+        Self { reader, cache }
+    }
+
+    /// Retrieve a reading copy of the store before starting the cache
+    #[must_use]
+    pub fn store(&self) -> Store<K> {
+        self.reader.clone()
+    }
+
+    /// Consume the stream and return a future that will run the duration of the program
+    ///
+    /// This should be awaited forever.
+    #[must_use]
+    pub fn run(self) -> BoxFuture<'static, ()> {
+        let stream = self.applies();
+        stream.for_each(|_| futures::future::ready(())).boxed()
+    }
+
+    /// Consumes the cache, runs the reflector, and returns an information stream of watch events (modified/added)
+    ///
+    /// Note that the returned stream is always reflected in the [`reader`](Cache::reader).
+    /// If you do not require a reader, prefer using a [`watcher`] directly.
+    pub fn applies(self) -> impl Stream<Item = K> + Send {
```

Comment:
This name feels pretty awkward.. `applied_objects` perhaps?

Reply:
Also, this feels like it's encouraging people to use reflectors even if they don't need them. I'd rather focus on how we can make `try_flatten_*` feel more ergonomic.

Reply:
I think the `utils` are fine by themselves (except possibly naming), I think the problem is that the users **have to** interact with them. Users generally just want a stream of objects and have a safe default. If we force people to use these `utils` on top of `watcher::Events`, then we have to teach everyone about the intricacies/failings of the watch api (we keep getting these questions), and I don't think we need to. The abstraction leaks out all the internals and it's too complicated for  people who just want to write something quick that just works.

The way I see it is that `watcher::Events` should be an implementation detail for `reflector` to do atomic cache transitions, and users should not be exposed to them unless they go looking for something advanced.